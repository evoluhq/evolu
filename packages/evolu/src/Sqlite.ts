import { Brand, Context, Effect, ReadonlyRecord } from "effect";

export interface Sqlite {
  readonly exec: (
    arg: string | QueryObject
  ) => Effect.Effect<never, never, ReadonlyArray<Row>>;

  readonly changes: Effect.Effect<never, never, number>;
}

export const Sqlite = Context.Tag<Sqlite>("evolu/Sqlite");

export interface QueryObject {
  readonly sql: string;
  readonly parameters: ReadonlyArray<Value>;
}

export type Value = null | string | number | Uint8Array;

export type Row = ReadonlyRecord.ReadonlyRecord<Value>;

export type Query = string & Brand.Brand<"Query">;

export const queryObjectToQuery = ({ sql, parameters }: QueryObject): Query =>
  JSON.stringify({ sql, parameters }) as Query;

export const queryObjectFromQuery = (s: Query): QueryObject =>
  JSON.parse(s) as QueryObject;
