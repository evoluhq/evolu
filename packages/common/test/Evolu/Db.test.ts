import { expect, test } from "vitest";
import { CallbackId } from "../../src/Callbacks.js";
import { createConsole } from "../../src/Console.js";
import {
  createDbWorkerForPlatform,
  DbSchema,
  DbWorker,
  DbWorkerOutput,
  DbWorkerPlatformDeps,
  getDbSnapshot,
} from "../../src/Evolu/Db.js";
import {
  Base64Url256,
  DbChange,
  idToBinaryId,
} from "../../src/Evolu/Protocol.js";
import { constVoid } from "../../src/Function.js";
import { wait } from "../../src/Promise.js";
import { getOrThrow } from "../../src/Result.js";
import { createSqlite, sql, Sqlite } from "../../src/Sqlite.js";
import {
  testCreateId,
  testCreateMnemonic,
  testCreateRandomBytesDep,
  testCreateSqliteDriver,
  testDbConfig,
  testNanoIdLib,
  testOwnerBinaryId,
  testRandom,
  testSimpleName,
  testTime,
} from "../_deps.js";
import { testTimestampsAsc } from "./_fixtures.js";

const createSimpleTestSchema = (): DbSchema => {
  return {
    tables: [
      {
        name: "testTable" as Base64Url256,
        columns: ["id" as Base64Url256, "name" as Base64Url256],
      },
      {
        name: "_testTable" as Base64Url256,
        columns: ["id" as Base64Url256, "name" as Base64Url256],
      },
    ],
    indexes: [],
  };
};

const createSqliteWithDbWorkerPlatformDeps = async (): Promise<
  [Sqlite, DbWorkerPlatformDeps]
> => {
  const sqliteDriver = await testCreateSqliteDriver(testSimpleName);
  const createSqliteDriver = () => Promise.resolve(sqliteDriver);
  const sqlite = getOrThrow(
    await createSqlite({ createSqliteDriver })(testSimpleName),
  );

  const deps: DbWorkerPlatformDeps = {
    createSqliteDriver,
    createSync: () => () => ({
      send: constVoid,
    }),
    console: createConsole(),
    time: testTime,
    random: testRandom,
    nanoIdLib: testNanoIdLib,
    createMnemonic: testCreateMnemonic,
    ...testCreateRandomBytesDep,
  };
  return [sqlite, deps];
};

const setupInitializedDbWorker = async (
  callback?: (db: DbWorker) => void,
): Promise<[Array<DbWorkerOutput>, Sqlite, DbWorker]> => {
  const [sqlite, deps] = await createSqliteWithDbWorkerPlatformDeps();
  const db = createDbWorkerForPlatform(deps);

  const dbWorkerOutput: Array<DbWorkerOutput> = [];
  db.onMessage((message) => dbWorkerOutput.push(message));

  // Execute callback before initialization if provided
  if (callback) {
    callback(db);
  }

  db.postMessage({
    type: "init",
    config: testDbConfig,
    dbSchema: createSimpleTestSchema(),
  });

  // async createSqliteDriver
  await wait(10);

  return [dbWorkerOutput, sqlite, db];
};

test("createDbWorker initializes correctly", async () => {
  const [dbWorkerOutput, sqlite] = await setupInitializedDbWorker();

  expect(dbWorkerOutput).toMatchSnapshot();
  expect(getDbSnapshot({ sqlite })).toMatchSnapshot();
});

test("mutations", async () => {
  const [dbWorkerOutput, sqlite, db] = await setupInitializedDbWorker();

  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [
      {
        id: testCreateId(),
        table: "testTable" as Base64Url256,
        values: { ["name" as Base64Url256]: "test" },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  expect(dbWorkerOutput).toMatchSnapshot();
  expect(getDbSnapshot({ sqlite }).tables).toMatchSnapshot();
});

test("mutate before init", async () => {
  const [dbWorkerOutput, sqlite] = await setupInitializedDbWorker((db) => {
    // This runs BEFORE init
    db.postMessage({
      type: "mutate",
      tabId: testCreateId(),
      changes: [
        {
          id: testCreateId(),
          table: "_testTable" as Base64Url256,
          values: { ["name" as Base64Url256]: "test" },
        },
      ],
      onCompleteIds: [],
      subscribedQueries: [],
    });
  });

  expect(dbWorkerOutput).toMatchSnapshot();
  expect(getDbSnapshot({ sqlite }).tables).toMatchSnapshot();
});

test("local mutation", async () => {
  const [dbWorkerOutput, sqlite, db] = await setupInitializedDbWorker();

  const change: DbChange = {
    id: testCreateId(),
    table: "_testTable" as Base64Url256,
    values: { ["name" as Base64Url256]: "test" },
  };

  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [change],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  expect(dbWorkerOutput).toMatchSnapshot();
  expect(getDbSnapshot({ sqlite }).tables).toMatchSnapshot();

  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [
      {
        ...change,
        values: {
          ...change.values,
          ["isDeleted" as Base64Url256]: 1,
        },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  expect(getDbSnapshot({ sqlite }).tables).toMatchSnapshot();
});

test("reset", async () => {
  const [, sqlite, db] = await setupInitializedDbWorker();

  db.postMessage({
    type: "reset",
    reload: false,
    onCompleteId: testNanoIdLib.nanoid() as CallbackId,
  });

  expect(getDbSnapshot({ sqlite })).toMatchSnapshot();
});

test("evolu_history unique index prevents duplicates", async () => {
  const [, sqlite] = await setupInitializedDbWorker();

  const ownerId = testOwnerBinaryId;
  const table = "testTable";
  const id = idToBinaryId(testCreateId());
  const column = "name";
  const value = "test value";
  const timestamp = testTimestampsAsc[0];

  // Manually insert the same record twice
  sqlite.exec(sql`
    insert into evolu_history
      ("ownerId", "table", "id", "column", "value", "timestamp")
    values
      (${ownerId}, ${table}, ${id}, ${column}, ${value}, ${timestamp})
    on conflict do nothing;
  `);
  sqlite.exec(sql`
    insert into evolu_history
      ("ownerId", "table", "id", "column", "value", "timestamp")
    values
      (${ownerId}, ${table}, ${id}, ${column}, ${value}, ${timestamp})
    on conflict do nothing;
  `);

  const count = getOrThrow(
    sqlite.exec<{ count: number }>(sql`
      select count(*) as count from evolu_history;
    `),
  );
  expect(count.rows[0].count).toBe(1);
});

test("timestamp ordering - newer mutations overwrite older ones", async () => {
  const [, sqlite, db] = await setupInitializedDbWorker();

  const recordId = testCreateId();

  // Create first mutation
  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [
      {
        id: recordId,
        table: "testTable" as Base64Url256,
        values: { ["name" as Base64Url256]: "first_value" },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  await wait(10);

  // Create second mutation on same record (will have newer timestamp)
  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [
      {
        id: recordId,
        table: "testTable" as Base64Url256,
        values: { ["name" as Base64Url256]: "second_value" },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  await wait(10);

  // Verify the app table has the latest value
  const finalResult = getOrThrow(
    sqlite.exec<{ name: string }>(sql`
      select name from testTable where id = ${recordId};
    `),
  );
  expect(finalResult.rows[0].name).toBe("second_value");

  // Verify both mutations are stored in history
  const historyCount = getOrThrow(
    sqlite.exec<{ count: number }>(sql`
      select count(*) as count
      from evolu_history
      where
        "table" = 'testTable'
        and "id" = ${idToBinaryId(recordId)}
        and "column" = 'name';
    `),
  );
  expect(historyCount.rows[0].count).toBe(2);
});

test("timestamp ordering - multiple columns update independently", async () => {
  const [, sqlite, db] = await setupInitializedDbWorker();

  const recordId = testCreateId();

  // Create first mutation that sets the name
  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [
      {
        id: recordId,
        table: "testTable" as Base64Url256,
        values: { ["name" as Base64Url256]: "original_name" },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  await wait(10);

  // Update the same record with a different value for name
  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [
      {
        id: recordId,
        table: "testTable" as Base64Url256,
        values: { ["name" as Base64Url256]: "updated_name" },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  await wait(10);

  // Verify the app table has the latest name value
  const finalResult = getOrThrow(
    sqlite.exec<{ name: string }>(sql`
      select name from testTable where id = ${recordId};
    `),
  );
  expect(finalResult.rows[0].name).toBe("updated_name");

  // Verify we have two entries in history for the name column
  const nameHistoryCount = getOrThrow(
    sqlite.exec<{ count: number }>(sql`
      select count(*) as count
      from evolu_history
      where
        "table" = 'testTable'
        and "id" = ${idToBinaryId(recordId)}
        and "column" = 'name';
    `),
  );
  expect(nameHistoryCount.rows[0].count).toBe(2);

  // Verify the values are stored in chronological order in history
  const historyValues = getOrThrow(
    sqlite.exec<{ value: string }>(sql`
      select value
      from evolu_history
      where
        "table" = 'testTable'
        and "id" = ${idToBinaryId(recordId)}
        and "column" = 'name'
      order by timestamp;
    `),
  );
  expect(historyValues.rows[0].value).toBe("original_name");
  expect(historyValues.rows[1].value).toBe("updated_name");
});

test("timestamp ordering - concurrent mutations on different records", async () => {
  const [, sqlite, db] = await setupInitializedDbWorker();

  const recordId1 = testCreateId();
  const recordId2 = testCreateId();

  // Create mutations on different records in quick succession
  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [
      {
        id: recordId1,
        table: "testTable" as Base64Url256,
        values: { ["name" as Base64Url256]: "record1_value" },
      },
      {
        id: recordId2,
        table: "testTable" as Base64Url256,
        values: { ["name" as Base64Url256]: "record2_value" },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  await wait(10);

  // Verify both records exist with correct values
  const allRecords = getOrThrow(
    sqlite.exec<{ id: string; name: string }>(sql`
      select id, name
      from testTable
      where id in (${recordId1}, ${recordId2})
      order by id;
    `),
  );
  expect(allRecords.rows).toHaveLength(2);

  const record1 = allRecords.rows.find((r) => r.id === recordId1);
  const record2 = allRecords.rows.find((r) => r.id === recordId2);

  expect(record1?.name).toBe("record1_value");
  expect(record2?.name).toBe("record2_value");

  // Verify both records have entries in history
  const totalHistoryCount = getOrThrow(
    sqlite.exec<{ count: number }>(sql`
      select count(*) as count
      from evolu_history
      where "table" = 'testTable' and "column" = 'name';
    `),
  );
  expect(totalHistoryCount.rows[0].count).toBe(2);
});

test("timestamp ordering - verify CRDT last-write-wins behavior", async () => {
  const [, sqlite, db] = await setupInitializedDbWorker();

  const recordId = testCreateId();

  // Create initial value
  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [
      {
        id: recordId,
        table: "testTable" as Base64Url256,
        values: { ["name" as Base64Url256]: "initial" },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  await wait(10);

  // Update multiple times rapidly to ensure different timestamps
  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [
      {
        id: recordId,
        table: "testTable" as Base64Url256,
        values: { ["name" as Base64Url256]: "second" },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  await wait(5);

  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [
      {
        id: recordId,
        table: "testTable" as Base64Url256,
        values: { ["name" as Base64Url256]: "third" },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  await wait(5);

  db.postMessage({
    type: "mutate",
    tabId: testCreateId(),
    changes: [
      {
        id: recordId,
        table: "testTable" as Base64Url256,
        values: { ["name" as Base64Url256]: "final" },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  await wait(10);

  // Verify app table has the final value (last write wins)
  const appTableResult = getOrThrow(
    sqlite.exec<{ name: string }>(sql`
      select name from testTable where id = ${recordId};
    `),
  );
  expect(appTableResult.rows[0].name).toBe("final");

  // Verify all mutations are preserved in history in timestamp order
  const historyResults = getOrThrow(
    sqlite.exec<{ value: string }>(sql`
      select value
      from evolu_history
      where
        "table" = 'testTable'
        and "id" = ${idToBinaryId(recordId)}
        and "column" = 'name'
      order by timestamp;
    `),
  );

  expect(historyResults.rows).toHaveLength(4);
  expect(historyResults.rows[0].value).toBe("initial");
  expect(historyResults.rows[1].value).toBe("second");
  expect(historyResults.rows[2].value).toBe("third");
  expect(historyResults.rows[3].value).toBe("final");

  // Verify that the app table always reflects the value with the highest timestamp
  const timestampResults = getOrThrow(
    sqlite.exec<{ value: string; timestamp: Uint8Array }>(sql`
      select value, timestamp
      from evolu_history
      where
        "table" = 'testTable'
        and "id" = ${idToBinaryId(recordId)}
        and "column" = 'name'
      order by timestamp desc
      limit 1;
    `),
  );

  expect(timestampResults.rows[0].value).toBe("final");
});
