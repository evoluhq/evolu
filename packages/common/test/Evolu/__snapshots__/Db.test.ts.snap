// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createDbWorker initializes correctly 1`] = `
[
  {
    "appOwner": {
      "encryptionKey": Uint8Array [
        173,
        165,
        204,
        225,
        74,
        240,
        76,
        99,
        46,
        18,
        104,
        4,
        0,
        136,
        188,
        132,
        99,
        57,
        151,
        4,
        147,
        44,
        59,
        91,
        61,
        162,
        116,
        243,
        186,
        98,
        25,
        4,
      ],
      "id": "MdVYFAxShUluuZKVWQfYL",
      "mnemonic": "knee easy fork attitude drink gloom head latin spider grab spy reason",
      "type": "AppOwner",
      "writeKey": Uint8Array [
        111,
        64,
        118,
        108,
        245,
        65,
        176,
        98,
        173,
        247,
        85,
        78,
        30,
        167,
        192,
        203,
      ],
    },
    "isFirst": true,
    "type": "onInit",
  },
]
`;

exports[`createDbWorker initializes correctly 2`] = `
{
  "schema": {
    "indexes": [
      {
        "name": "evolu_history_ownerId_timestamp",
        "sql": "create index evolu_history_ownerId_timestamp on evolu_history (
          "ownerId",
          "timestamp"
        )",
      },
      {
        "name": "evolu_history_ownerId_table_id_column_timestampDesc",
        "sql": "create unique index evolu_history_ownerId_table_id_column_timestampDesc on evolu_history (
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp" desc
        )",
      },
      {
        "name": "evolu_timestamp_index",
        "sql": "create index evolu_timestamp_index on evolu_timestamp (
        "ownerId",
        "l",
        "t",
        "h1",
        "h2",
        "c"
      )",
      },
    ],
    "tables": [
      {
        "columns": [
          "protocolVersion",
        ],
        "name": "evolu_version",
      },
      {
        "columns": [
          "clock",
          "appOwnerId",
          "appOwnerEncryptionKey",
          "appOwnerWriteKey",
          "appOwnerMnemonic",
        ],
        "name": "evolu_config",
      },
      {
        "columns": [
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp",
          "value",
        ],
        "name": "evolu_history",
      },
      {
        "columns": [
          "id",
          "name",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "testTable",
      },
      {
        "columns": [
          "id",
          "name",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "_testTable",
      },
      {
        "columns": [
          "ownerId",
          "t",
          "h1",
          "h2",
          "c",
          "l",
        ],
        "name": "evolu_timestamp",
      },
    ],
  },
  "tables": [
    {
      "name": "evolu_version",
      "rows": [
        {
          "protocolVersion": 0,
        },
      ],
    },
    {
      "name": "evolu_config",
      "rows": [
        {
          "appOwnerEncryptionKey": {
            "data": [
              173,
              165,
              204,
              225,
              74,
              240,
              76,
              99,
              46,
              18,
              104,
              4,
              0,
              136,
              188,
              132,
              99,
              57,
              151,
              4,
              147,
              44,
              59,
              91,
              61,
              162,
              116,
              243,
              186,
              98,
              25,
              4,
            ],
            "type": "Buffer",
          },
          "appOwnerId": "MdVYFAxShUluuZKVWQfYL",
          "appOwnerMnemonic": "knee easy fork attitude drink gloom head latin spider grab spy reason",
          "appOwnerWriteKey": {
            "data": [
              111,
              64,
              118,
              108,
              245,
              65,
              176,
              98,
              173,
              247,
              85,
              78,
              30,
              167,
              192,
              203,
            ],
            "type": "Buffer",
          },
          "clock": "1970-01-01T00:00:00.000Z-0000-40aaf4d24f2982ba",
        },
      ],
    },
    {
      "name": "evolu_history",
      "rows": [],
    },
    {
      "name": "testTable",
      "rows": [],
    },
    {
      "name": "_testTable",
      "rows": [],
    },
    {
      "name": "evolu_timestamp",
      "rows": [],
    },
  ],
}
`;

exports[`local mutation 1`] = `
[
  {
    "appOwner": {
      "encryptionKey": Uint8Array [
        173,
        165,
        204,
        225,
        74,
        240,
        76,
        99,
        46,
        18,
        104,
        4,
        0,
        136,
        188,
        132,
        99,
        57,
        151,
        4,
        147,
        44,
        59,
        91,
        61,
        162,
        116,
        243,
        186,
        98,
        25,
        4,
      ],
      "id": "MdVYFAxShUluuZKVWQfYL",
      "mnemonic": "knee easy fork attitude drink gloom head latin spider grab spy reason",
      "type": "AppOwner",
      "writeKey": Uint8Array [
        111,
        64,
        118,
        108,
        245,
        65,
        176,
        98,
        173,
        247,
        85,
        78,
        30,
        167,
        192,
        203,
      ],
    },
    "isFirst": true,
    "type": "onInit",
  },
  {
    "onCompleteIds": [],
    "patches": [],
    "tabId": "c3jir3c4AAUWg7IbJo2Xp",
    "type": "onChange",
  },
  {
    "tabId": "c3jir3c4AAUWg7IbJo2Xp",
    "type": "onReceive",
  },
]
`;

exports[`local mutation 2`] = `
[
  {
    "name": "evolu_version",
    "rows": [
      {
        "protocolVersion": 0,
      },
    ],
  },
  {
    "name": "evolu_config",
    "rows": [
      {
        "appOwnerEncryptionKey": {
          "data": [
            173,
            165,
            204,
            225,
            74,
            240,
            76,
            99,
            46,
            18,
            104,
            4,
            0,
            136,
            188,
            132,
            99,
            57,
            151,
            4,
            147,
            44,
            59,
            91,
            61,
            162,
            116,
            243,
            186,
            98,
            25,
            4,
          ],
          "type": "Buffer",
        },
        "appOwnerId": "MdVYFAxShUluuZKVWQfYL",
        "appOwnerMnemonic": "knee easy fork attitude drink gloom head latin spider grab spy reason",
        "appOwnerWriteKey": {
          "data": [
            111,
            64,
            118,
            108,
            245,
            65,
            176,
            98,
            173,
            247,
            85,
            78,
            30,
            167,
            192,
            203,
          ],
          "type": "Buffer",
        },
        "clock": "1970-01-01T00:00:00.000Z-0000-7cd5aa540db883b2",
      },
    ],
  },
  {
    "name": "evolu_history",
    "rows": [],
  },
  {
    "name": "testTable",
    "rows": [],
  },
  {
    "name": "_testTable",
    "rows": [
      {
        "createdAt": "1970-01-01T00:00:00.002Z",
        "id": "gFeyTa9pE7z9KY-wzYDMk",
        "isDeleted": null,
        "name": "test",
        "updatedAt": "1970-01-01T00:00:00.002Z",
      },
    ],
  },
  {
    "name": "evolu_timestamp",
    "rows": [],
  },
]
`;

exports[`local mutation 3`] = `
[
  {
    "name": "evolu_version",
    "rows": [
      {
        "protocolVersion": 0,
      },
    ],
  },
  {
    "name": "evolu_config",
    "rows": [
      {
        "appOwnerEncryptionKey": {
          "data": [
            173,
            165,
            204,
            225,
            74,
            240,
            76,
            99,
            46,
            18,
            104,
            4,
            0,
            136,
            188,
            132,
            99,
            57,
            151,
            4,
            147,
            44,
            59,
            91,
            61,
            162,
            116,
            243,
            186,
            98,
            25,
            4,
          ],
          "type": "Buffer",
        },
        "appOwnerId": "MdVYFAxShUluuZKVWQfYL",
        "appOwnerMnemonic": "knee easy fork attitude drink gloom head latin spider grab spy reason",
        "appOwnerWriteKey": {
          "data": [
            111,
            64,
            118,
            108,
            245,
            65,
            176,
            98,
            173,
            247,
            85,
            78,
            30,
            167,
            192,
            203,
          ],
          "type": "Buffer",
        },
        "clock": "1970-01-01T00:00:00.000Z-0000-7cd5aa540db883b2",
      },
    ],
  },
  {
    "name": "evolu_history",
    "rows": [],
  },
  {
    "name": "testTable",
    "rows": [],
  },
  {
    "name": "_testTable",
    "rows": [],
  },
  {
    "name": "evolu_timestamp",
    "rows": [],
  },
]
`;

exports[`mutate before init 1`] = `
[
  {
    "appOwner": {
      "encryptionKey": Uint8Array [
        173,
        165,
        204,
        225,
        74,
        240,
        76,
        99,
        46,
        18,
        104,
        4,
        0,
        136,
        188,
        132,
        99,
        57,
        151,
        4,
        147,
        44,
        59,
        91,
        61,
        162,
        116,
        243,
        186,
        98,
        25,
        4,
      ],
      "id": "MdVYFAxShUluuZKVWQfYL",
      "mnemonic": "knee easy fork attitude drink gloom head latin spider grab spy reason",
      "type": "AppOwner",
      "writeKey": Uint8Array [
        111,
        64,
        118,
        108,
        245,
        65,
        176,
        98,
        173,
        247,
        85,
        78,
        30,
        167,
        192,
        203,
      ],
    },
    "isFirst": true,
    "type": "onInit",
  },
  {
    "onCompleteIds": [],
    "patches": [],
    "tabId": "Df11mF2BAPRJEKIftYOy3",
    "type": "onChange",
  },
  {
    "tabId": "Df11mF2BAPRJEKIftYOy3",
    "type": "onReceive",
  },
]
`;

exports[`mutate before init 2`] = `
[
  {
    "name": "evolu_version",
    "rows": [
      {
        "protocolVersion": 0,
      },
    ],
  },
  {
    "name": "evolu_config",
    "rows": [
      {
        "appOwnerEncryptionKey": {
          "data": [
            173,
            165,
            204,
            225,
            74,
            240,
            76,
            99,
            46,
            18,
            104,
            4,
            0,
            136,
            188,
            132,
            99,
            57,
            151,
            4,
            147,
            44,
            59,
            91,
            61,
            162,
            116,
            243,
            186,
            98,
            25,
            4,
          ],
          "type": "Buffer",
        },
        "appOwnerId": "MdVYFAxShUluuZKVWQfYL",
        "appOwnerMnemonic": "knee easy fork attitude drink gloom head latin spider grab spy reason",
        "appOwnerWriteKey": {
          "data": [
            111,
            64,
            118,
            108,
            245,
            65,
            176,
            98,
            173,
            247,
            85,
            78,
            30,
            167,
            192,
            203,
          ],
          "type": "Buffer",
        },
        "clock": "1970-01-01T00:00:00.000Z-0000-41123f05b8442d71",
      },
    ],
  },
  {
    "name": "evolu_history",
    "rows": [],
  },
  {
    "name": "testTable",
    "rows": [],
  },
  {
    "name": "_testTable",
    "rows": [
      {
        "createdAt": "1970-01-01T00:00:00.001Z",
        "id": "ms4ropUP_dGi2jCaTw-bM",
        "isDeleted": null,
        "name": "test",
        "updatedAt": "1970-01-01T00:00:00.001Z",
      },
    ],
  },
  {
    "name": "evolu_timestamp",
    "rows": [],
  },
]
`;

exports[`mutations 1`] = `
[
  {
    "appOwner": {
      "encryptionKey": Uint8Array [
        173,
        165,
        204,
        225,
        74,
        240,
        76,
        99,
        46,
        18,
        104,
        4,
        0,
        136,
        188,
        132,
        99,
        57,
        151,
        4,
        147,
        44,
        59,
        91,
        61,
        162,
        116,
        243,
        186,
        98,
        25,
        4,
      ],
      "id": "MdVYFAxShUluuZKVWQfYL",
      "mnemonic": "knee easy fork attitude drink gloom head latin spider grab spy reason",
      "type": "AppOwner",
      "writeKey": Uint8Array [
        111,
        64,
        118,
        108,
        245,
        65,
        176,
        98,
        173,
        247,
        85,
        78,
        30,
        167,
        192,
        203,
      ],
    },
    "isFirst": true,
    "type": "onInit",
  },
  {
    "onCompleteIds": [],
    "patches": [],
    "tabId": "v62hCHsxOuqK4AbttUxKh",
    "type": "onChange",
  },
  {
    "tabId": "v62hCHsxOuqK4AbttUxKh",
    "type": "onReceive",
  },
]
`;

exports[`mutations 2`] = `
[
  {
    "name": "evolu_version",
    "rows": [
      {
        "protocolVersion": 0,
      },
    ],
  },
  {
    "name": "evolu_config",
    "rows": [
      {
        "appOwnerEncryptionKey": {
          "data": [
            173,
            165,
            204,
            225,
            74,
            240,
            76,
            99,
            46,
            18,
            104,
            4,
            0,
            136,
            188,
            132,
            99,
            57,
            151,
            4,
            147,
            44,
            59,
            91,
            61,
            162,
            116,
            243,
            186,
            98,
            25,
            4,
          ],
          "type": "Buffer",
        },
        "appOwnerId": "MdVYFAxShUluuZKVWQfYL",
        "appOwnerMnemonic": "knee easy fork attitude drink gloom head latin spider grab spy reason",
        "appOwnerWriteKey": {
          "data": [
            111,
            64,
            118,
            108,
            245,
            65,
            176,
            98,
            173,
            247,
            85,
            78,
            30,
            167,
            192,
            203,
          ],
          "type": "Buffer",
        },
        "clock": "1970-01-01T00:00:00.000Z-0001-a8d0151b0fb0b791",
      },
    ],
  },
  {
    "name": "evolu_history",
    "rows": [
      {
        "column": "name",
        "id": {
          "data": [
            222,
            71,
            117,
            30,
            223,
            109,
            157,
            93,
            143,
            80,
            64,
            84,
            254,
            164,
            40,
            176,
          ],
          "type": "Buffer",
        },
        "ownerId": {
          "data": [
            128,
            87,
            31,
            173,
            149,
            230,
            206,
            93,
            128,
            2,
            246,
            220,
            162,
            236,
            95,
            168,
          ],
          "type": "Buffer",
        },
        "table": "testTable",
        "timestamp": {
          "data": [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            1,
            168,
            208,
            21,
            27,
            15,
            176,
            183,
            145,
          ],
          "type": "Buffer",
        },
        "value": "test",
      },
    ],
  },
  {
    "name": "testTable",
    "rows": [
      {
        "createdAt": null,
        "id": "qBEkmGiGH5l37ns7tL4W_",
        "isDeleted": null,
        "name": "test",
        "updatedAt": "1970-01-01T00:00:00.000Z",
      },
    ],
  },
  {
    "name": "_testTable",
    "rows": [],
  },
  {
    "name": "evolu_timestamp",
    "rows": [
      {
        "c": 1,
        "h1": 255490510088691,
        "h2": 18088005276510,
        "l": 2,
        "ownerId": {
          "data": [
            128,
            87,
            31,
            173,
            149,
            230,
            206,
            93,
            128,
            2,
            246,
            220,
            162,
            236,
            95,
            168,
          ],
          "type": "Buffer",
        },
        "t": {
          "data": [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            1,
            168,
            208,
            21,
            27,
            15,
            176,
            183,
            145,
          ],
          "type": "Buffer",
        },
      },
    ],
  },
]
`;

exports[`reset 1`] = `
{
  "schema": {
    "indexes": [],
    "tables": [],
  },
  "tables": [],
}
`;
